bundles/org.openhab.binding.mpd/src/main/java/org/openhab/binding/mpd/internal/protocol/MPDConnectionThread.java

```diff
@@ -0,0 +1,298 @@
+        logger.debug("insert command '{}' at position {}", command.getCommand(), position);
+        int index = position;
+        synchronized (pendingCommands) {
+            if (index < 0) {
+                index = pendingCommands.size();
+            }
+            pendingCommands.add(index, command);
+            sendNoIdleIfInIdle();
+        }
+    }
+
+    private void updateThingStatus(ThingStatus status, ThingStatusDetail statusDetail, @Nullable String description) {
+        if (!disposed.get()) {
+            listener.updateThingStatus(status, statusDetail, description);
+        }
+    }
+
+    private void sendNoIdleIfInIdle() {
+        if (isInIdle.compareAndSet(true, false)) {
+            try {
+                sendCommand(new MPDCommand("noidle"));
+            } catch (IOException e) {
+                logger.debug("sendCommand(noidle) failed");
+            }
+        }
+    }
+
+    private void establishConnection() throws IOException, MPDException {
+        openSocket();
+
+        MPDCommand currentCommand = new MPDCommand("connect");
+        MPDResponse response = readResponse(currentCommand);
+
+        if (!response.isOk()) {
+            throw new MPDException("could not connect");
+        }
+
+        if (!password.isEmpty()) {
+            currentCommand = new MPDCommand("password", password);
+            sendCommand(currentCommand);
+            response = readResponse(currentCommand);
+            if (!response.isOk()) {
+                throw new MPDException("could not authenticate");
+            }
+        }
+    }
+
+    private void openSocket() throws IOException {
+        logger.debug("opening connection to {} port {}", address, port);
+        Socket socket = new Socket(address, port);
+
+        inputStreamReader = new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8);
+        reader = new BufferedReader(inputStreamReader);
+        writer = new DataOutputStream(socket.getOutputStream());
+
+        this.socket = socket;
+    }
+
+    private void processPendingCommands() throws IOException {
+        MPDCommand currentCommand;
+
+        while (!disposed.get()) {
+            synchronized (pendingCommands) {
+                if (!pendingCommands.isEmpty()) {
+                    currentCommand = pendingCommands.remove(0);
+                } else {
+                    currentCommand = new MPDCommand("idle");
+                }
+
+                sendCommand(currentCommand);
+                if (currentCommand.getCommand().equals("idle")) {
+                    isInIdle.set(true);
+                }
+            }
+
+            MPDResponse response = readResponse(currentCommand);
+            if (!response.isOk()) {
+                insertCommand(new MPDCommand("clearerror"), 0);
+            }
+            listener.onResponse(response);
+        }
+    }
+
+    private void closeSocket() {
+        logger.debug("Closing socket");
```

Comment:
Do you really need to close all of them? Will closing some of them not take a long the rest. E.g closing both the `reader` and `inputStreamReader` is AFAIK not needed.

A quick [peak in the javadoc](https://docs.oracle.com/javase/7/docs/api/java/net/Socket.html#close()) suggested that closing the socket will do all the work at one.

Reply:
I have realized that the writer (DataOutputStream) is not necessary, since one can use socket.getOutputStream() directly.
@fwolter has suggested in https://github.com/openhab/openhab-addons/pull/7870#discussion_r451814387 to also close the InputStreamReader, which is in contrast to your suggestion. From the mentioned javadoc it is clear that the InputStream and the OutputStream will also be closed, when the socket is closed. I do not close these streams explicitly.
However, I would at least close the BufferedReader. But it does not harm, if I also close the InputStreamReader. I prefer to close too much than too little. wdyt?

Reply:
I'm okay with keeping it like this. But there is a reason they mention it. Sometimes closing twice could lead to errors/exceptions and when not explicitly documented I would rather not.